Clarify 'colr' box and color info in codecs string

wantehchang opened the pull request:
Fix https://github.com/AOMediaCodec/av1-isobmff/issues/126 and
https://github.com/AOMediaCodec/av1-isobmff/issues/129.

cconcolato commented:
The changes in the first part (#126) implement what was suggested in the issue. We realize this is a normative change with potential backwards compatibility issues. We need to check with implementations.

For the codecs parameter changes (#129), we would prefer:
- bullet point lists and indentation for readability
- having the range conditions separate as they do not dependent on the color present flag
This is also a normative change but fixing a case where there was conflicting information between the `codecs` parameter and the bitstream. We need careful review.

In both cases, we should document before and after behavior.

wantehchang commented:
Here is a screen shot of the first change in this pull request:
![Screenshot from 2022-06-13 09-53-19](https://user-images.githubusercontent.com/16024799/173405167-742a43b3-b257-4611-99a0-e5d05e185b74.png)

Here is a screen shot of the second change in this pull request. My change ends with the paragraph "The videoFullRangeFlag is represented by a single digit." The screen shot includes a few paragraphs below and the table of default values as context.

![Screenshot from 2022-06-13 09-51-48](https://user-images.githubusercontent.com/16024799/173405206-8a2915e5-1fae-4cdf-b250-242ae15e80bb.png)

cconcolato commented:
We should update the PR again to cover the case of monochrome and chromaSubsampling. They cannot be present if the following fields are omitted (and vice versa).

cconcolato commented:
The codecs parameter is also allowed to omit the trailing values if the colr box or the sequence header OBU contains the default values.

cconcolato commented:
We also need to document the before and after cases, showing possible backwards compatibility changes.

wantehchang commented:
The new version of the pull request has addressed all the review comments except the following:
- The value of the videoFullRangeFlag is not specified separately. This results in a little duplicate text but I think overall this is clearer.
- I didn't document the before and after cases, showing possible backwards compatibility changes.

podborski commented:
Question, should `videoFullRangeFlag` be `color_range` here as well? (similar fix as in HDR10+ pull request?) What about other terms?

wantehchang commented:
Screenshot of the top half of Section 5:

![image](https://user-images.githubusercontent.com/16024799/201732369-131d4c63-f4fe-4257-806a-e86fb048c068.png)

wantehchang commented:
Screenshot of the bottom half of Section 5:

![image](https://user-images.githubusercontent.com/16024799/201732646-b9d4fe29-3f86-4cc2-acd0-132d07f4deba.png)

cconcolato commented:
We should split the PR into 2:
* one clarifying the `colr` box and Config OBU part
* one clarifying the `codecs` string

cconcolato commented:
For the `colr`, I suggest we clarify along the lines of [this document](https://docs.google.com/document/d/1xJtqj8VnC5bZqbVnDvDTNAn3buD9BZV8TUhuDFN-EDY/edit#)

cconcolato commented:
Regarding the `codecs` string part, the added sentence seems wrong:
>  If any of the colorPrimaries, transferCharacteristics, and matrixCoefficients parameter values equals 02 (unspecified), it SHALL be set to the default value below  .

It does not enable 2/2/2 for DV for example. Also the "it" is not clear. Simply deleting the sentence is good, or moving it into the "if `colr` box is absent.
We could add a note that says: if you explicitly want 2/2/2 in the `codecs` string, you should use a `colr` box with these 2/2/2 values.

wantehchang commented:
I wrote the pull request https://github.com/AOMediaCodec/av1-isobmff/pull/167 to clarify the `colr` box.

cconcolato commented:
Some notes on the PR:
- we note that there are colr boxes with other types than nclx (e.g. nclc, or prof). The PR considers that anything else that nclx requires going into the bitstream. Is that what we want? Probably.
- There are different tools processing codecs strings: MPD generator, players, ... We wonder how complex the generation of the codecs string should be for MPD generator (assuming more work can be done by encoder/packager, when creating the MP4 file from the OBU stream). The alternative would be to never read the sequence header? How many files would this invalidate?
- Maybe we should add a note that if you use a color space that is not registered in CICP, if you don't explicitly code 2/2/2 either in the sequence header (with color description present flag = 1) or in the colr box, you will get a codecs string that declares BT 709. Writers should act accordingly.

podborski commented:
Maybe one more thing to consider. The [HLS spec.](https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#section-4.4.6.2) also defines a `VIDEO-RANGE` as

<details>
<summary>VIDEO-RANGE definition</summary>
VIDEO-RANGE

      The value is an enumerated-string; valid strings are SDR, HLG and
      PQ.

      The value MUST be SDR if the video in the Variant Stream is
      encoded using one of the following reference opto-electronic
      transfer characteristic functions specified by the
      TransferCharacteristics code point: [[CICP](https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#ref-CICP)] 1, 6, 13, 14, 15.  Note
      that different TransferCharacteristics code points can use the
      same transfer function.

     The value MUST be HLG if the video in the Variant Stream is
      encoded using a reference opto-electronic transfer characteristic
      function specified by the TransferCharacteristics code point 18,
      or consists of such video mixed with video qualifying as SDR (see
      above).

      The value MUST be PQ if the video in the Variant Stream is encoded
      using a reference opto-electronic transfer characteristic function
      specified by the TransferCharacteristics code point 16, or
      consists of such video mixed with video qualifying as SDR or HLG
      (see above).

      This attribute is OPTIONAL.  Its absence implies a value of SDR.
      Clients that do not recognize the attribute value SHOULD NOT
      select the Variant Stream.
</details>

I think it would be useful if we at least add a note, mentioning that care needs to be taken as the  `VIDEO-RANGE` attribute in HLS groups certain transfer characteristics CICP values into 3 buckets (SDR,HLG,PQ) and codecs string is signaling similar information.

cconcolato commented:
We could add a note that says:
NOTE: when the `colr` box is present, it simplifies the task of the `codecs` parameter generator, and is encouraged.

wantehchang commented:
I added the NOTE that Cyril suggested. I also move the common text about `color_range` to a separate bullet point.

wantehchang commented:
Cyril,

I have some questions about your comment on Jul 10.

> * we note that there are colr boxes with other types than nclx (e.g. nclc, or prof).

I assume "nclc" is a typo for "'rICC".

> The PR considers that anything else that nclx requires going into the bitstream. Is that what we want? Probably.

Could you clarify what you meant by "anything else that nclx requires going into the bitstream"? I wonder if you meant to say "anything other than nclx requires going into the bitstream".

By "going into the bitstream", you meant going into the Sequence Header OBU, right?

> * Maybe we should add a note that if you use a color space that is not registered in CICP, if you don't explicitly code 2/2/2 either in the sequence header (with color description present flag = 1) or in the colr box, you will get a codecs string that declares BT 709. Writers should act accordingly.

Would you still like me to add this note?

Thanks.

wantehchang commented:
Dimitri wrote:

> Maybe one more thing to consider. The [HLS spec.](https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#section-4.4.6.2) also defines a `VIDEO-RANGE` [...snipped...]
> 
> I think it would be useful if we at least add a note, mentioning that care needs to be taken as the `VIDEO-RANGE` attribute in HLS groups certain transfer characteristics CICP values into 3 buckets (SDR,HLG,PQ) and codecs string is signaling similar information.

Should I add the note that Dimitri suggested?

The HLS spec that Dimitri mentioned is still an Internet Draft. It seems premature to mention it.

cconcolato commented:
> I assume "nclc" is a typo for "'rICC".

No. This was an example. `nclc` exists, typically in QuickTime files.

cconcolato commented:
> Could you clarify what you meant by "anything else that nclx requires going into the bitstream"? I wonder if you meant to say "anything other than nclx requires going into the bitstream".

Yes, that's what I meant.

> By "going into the bitstream", you meant going into the Sequence Header OBU, right?

Yes.

cconcolato commented:
> Would you still like me to add this note?

Yes, that couldn't hurt.

cconcolato commented:
> Should I add the note that Dimitri suggested?

No. We said that Dimitri will open a separate issue to discuss HLS. Let's not overload this PR too much.

wantehchang commented:
OK, I have addressed all the comments on this pull request.

cconcolato commented on index.bs:
```diff
@@ -635,7 +635,20 @@ All fields following the sample entry 4CC are expressed as double digit decimals
+
+NOTE: It is encouraged that a 'colr' box be present. The 'colr' box simplifies the task of the codecs parameter generator.
+
+NOTE: If you use a color space that is not registered in CICP and you don't explicitly code 2/2/2 either in the 'colr' box or in the [=Sequence Header OBU=] (with color_description_present_flag set to 1), you will get a codecs parameter string that declares ITU-R BT.709 (1/1/1). Writers should act accordingly.
```
We need an informative reference to CICP. ISO/IEC 23091-2 | ITU.T H.273, with a link to the ITU website (free download)

cconcolato replied:
Rephrase the NOTE, avoiding the use of "you":
```suggestion
NOTE: If the encoded video uses a color space that is not registered in [[CICP]] and the file doesn't explicitly code 2/2/2 either in the 'colr' box or in the [=Sequence Header OBU=] (with color_description_present_flag set to 1), the derived codecs parameter string will declare ITU-R BT.709 (1/1/1). To avoid this behavior, a 'colr' box should be used.
```

cconcolato commented:
Unless objections in the coming days I will merge this PR. Thank you @wantehchang

rafaelcaricio commented on index.bs:
```diff
@@ -635,7 +641,20 @@ All fields following the sample entry 4CC are expressed as double digit decimals
+
+The videoFullRangeFlag is represented by a single digit.
+
+<assert>If the codecs parameter string ends with ".0.110.01.01.01.0" (containing all the default values below), that trailing part of the string SHOULD be omitted</assert>.
```
Maybe this could be changed from SHOULD to MAY? I think it would be an implementers choice to include it or not.

cconcolato replied:
Thanks for the comment. It's a SHOULD rather than a MAY because it is inefficient to include all the default values. But in the end, it's an implementer's choice. The SHOULD just gives more indication as to what needs to be done than a MAY. Unless we have a use case that needs to include the default values, I think we should keep the SHOULD.

ArthurGrizzle commented:
No

ArthurGrizzle commented:
Yes

cconcolato commented:
> Yes
> No

What are you commenting on?
